Checkpointer

jdrotleff opened the pull request:
# Checkpointer

A class that can be used to automatically save the current Neural Network if the reward or its running average meet some chosen criteria.
To use it, pass a checkpoint dictionary to the Trainer, in which you define your wanted save-criteria.

Currently, these options are implemented.

- Regular Backup: Saves a copy of the network every ```n``` episodes.
- Goal Backup: Saves a copy of the network as soon as a desired reward threshold is
reached. The simulation can possibly be stopped after the goal is reached, continued for a defined # of episodes or until it would end normally. 
- Best Model: Always saves a copy of the network, if the mean reward of the last ```window_width```
episodes is greater than before. The next best reward can be increased as a percentage. 
- Backup Model: Saves a copy of the network, if the mean reward of the last ```window_width``` episodes
stops increasing. This could possibly save models, that still did some learning while the
agents were successful. Alternatively, the backup model can be compared to the best
model to analyze the phenomenom of forgetting.

For every episode each criterion is checked. If at least one is fulfilled, the model gets exported. The name of the exported model contains the criteria that are met.

An example application of this class can look like this:
```
checkpointing_params = {'window_width' : 10,
                        'DO_GOAL_MODEL' : True,
                        'required_reward' : 250,
                        'DO_GOAL_BREAK' : True,
                        'running_out_length' : 25,
                        'DO_BEST_MODEL' : True,
                        'min_reward' : 250,
                        'increase_factor' : 1.05,
                        'DO_BACKUP_MODEL' : True,
                        'min_backup_reward': 250,
                        'DO_REGULAR' : True,
                        'save_models_intervall' : 25,
                        }

rl_trainer = srl.trainers.EpisodicTrainer(
    [agent], 
    checkpointing_params
)
rewards = rl_trainer.perform_rl_training(...)
```
The chosen parameters depend highly on the simulation performed.

For a more detailled explanation check the docstrings in trainer.py.

<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

- **New Features**
  - Built-in training checkpointing with multiple strategies (regular interval, best-reward, goal-based, backup).
  - Trainers export checkpoints during episodes and can early-stop when a reward goal is reached.
  - Trainer accepts configurable checkpointers and auto-selects checkpoint path.
  - Checkpointers exposed as a public module for easy access.

- **Refactor**
  - Episode rewards stored as a fixed-size array; added logging hooks and per-episode checkpoint integration.

- **Tests**
  - Added unit tests covering all checkpointer strategies, initialization, and trigger/stop behavior.
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

jdrotleff committed:
Added Checkpointer Class

This allows to check the reward for certain criteria to eventually save the current network.

jdrotleff committed:
Added checkpointing to Episodic Trainer

jdrotleff committed:
Added checkpointing to Continuous Trainer

jdrotleff committed:
Merge branch 'SwarmRL:master' into Checkpointer

pre-commit-ci-lite[bot] committed:
[pre-commit.ci lite] apply automatic fixes

jdrotleff committed:
Fixed Formatting

jdrotleff committed:
Fixed Formatting

pre-commit-ci-lite[bot] committed:
[pre-commit.ci lite] apply automatic fixes

jdrotleff committed:
Fixed Formatting

pre-commit-ci-lite[bot] committed:
[pre-commit.ci lite] apply automatic fixes

jdrotleff committed:
Fix undefined variable in case of no checkpointer

jdrotleff committed:
Fixed list/ndarray problem for passing the tests

jdrotleff committed:
Fix for semi-episodic case

jdrotleff committed:
Formatting error

SamTov commented:
I am more a fan of having a generic base class for checkpointing and then having child classes for the different kinds of checkpointing. This avoids having huge parameter dictionaries. We can then allow users to pass multiple checkpointers in a list.

SamTov committed:
Merge branch 'master' into Checkpointer

jdrotleff committed:
Changed Checkpointer into BaseClass/SubClass structure

pre-commit-ci-lite[bot] committed:
[pre-commit.ci lite] apply automatic fixes

jpd-de commented:
I changed the structure of the checkpointers to a base class/child class structure.


To use it, create Checkpointers of your choice and pass them to the Trainer.

Currently, these options are implemented.

- RegularCheckpointer: Saves a copy of the network every ```save_interval``` episodes.
- GoalCheckpointer: Saves a copy of the network as soon as a desired reward threshold is
reached. The simulation can possibly be stopped after the goal is reached, continued for a defined # of episodes or until it would end normally. 
- BestRewardCheckpointer: Always saves a copy of the network, if the mean reward of the last ```window_width```
episodes is greater than before. The next best reward can be increased as a percentage. 
- BackupCheckpointer: Saves a copy of the network, if the mean reward of the last ```window_width``` episodes
stops increasing. This could possibly save models, that still did some learning while the
agents were successful. Alternatively, the backup model can be compared to the best
model to analyze the phenomenom of forgetting.

General parameters:
```window_width```: How many episodes should be considered for the running reward average?
GoalCheckpointer parameters:
```running_out_length```: How many episodes should the simulation continue, as the goal is reached.
```required_reward```: Sets the wanted reward.
BestRewardCheckpointer and BackupCheckpointer parameters.
```increase_factor```: The factor the reward should be greater than the previous maximum reward.
```wait_time```: Sets the # of episodes that will not be saved after a best model was found. This prevents a huge amount of saves of the same model.


For every episode each criterion is checked. If at least one is fulfilled, the model gets exported. The name string of the exported model contains the criteria that are met.

An example use of all implemented methods could look like this:
```
from swarmrl.checkpointers import GoalCheckpointer, BestRewardCheckpointer, BackupCheckpointer, RegularCheckpointer

Goal_Checkpointer = GoalCheckpointer(required_reward=250, 
                                            running_out_length=25, 
                                            window_width=10,
                                            do_goal_break=True)
Best_Checkpointer = BestRewardCheckpointer(min_reward=250,
                                            increase_factor=1.05,
                                            window_width=10)
Backup_Checkpointer = BackupCheckpointer(min_backup_reward=250,
                                            wait_time=10,
                                            window_width=10)
Regular_Checkpointer = RegularCheckpointer(save_interval=25)

rl_trainer = srl.trainers.EpisodicTrainer(
    [agent], 
    [Goal_Checkpointer, Backup_Checkpointer, Regular_Checkpointer]
    )
```
Note that the chosen parameters depend highly on the simulation performed and your specific task.

SamTov commented on swarmrl/checkpointers/__init__.py:
```diff
@@ -0,0 +1,17 @@
+
+from swarmrl.checkpointers.backup_checkpointer import BackupCheckpointer
+from swarmrl.checkpointers.best_checkpointer import BestRewardCheckpointer
+from swarmrl.checkpointers.checkpointer import Checkpointer
```
This will seem pedantic but can we change `Checkpointer` to `BaseCheckpointer` to be very clear that it is a parent class.

jpd-de replied:
Sure! Makes it clearer.

SamTov commented on swarmrl/checkpointers/backup_checkpointer.py:
```diff
@@ -0,0 +1,54 @@
+
+class BackupCheckpointer(Checkpointer):
+    """
+    Checkpointer that saves the model if a new best reward is achieved.
```
Is this correct?

SamTov commented on swarmrl/checkpointers/best_checkpointer.py:
```diff
@@ -0,0 +1,64 @@
+    """
+
+    def __init__(
+        self, min_reward=250, increase_factor=1.05, window_width=30, wait_time=10
```
Can you add some doc string for this function to introduce each argument?

jpd-de replied:
Done.

SamTov commented on swarmrl/trainers/episodic_trainer.py:
```diff
@@ -65,7 +69,7 @@ def perform_rl_training(
         """
         killed = False
-        rewards = [0.0]
+        rewards = np.zeros(n_episodes)
```
Nice addition.

jpd-de replied:
Thanks!

SamTov reviewed (changes requested):
In general, I like it. I would like to see some more comprehensive documentation as well as some changes to the structure of a few checkpointers and names. Please take a look at the comments.

SamTov committed:
Merge branch 'master' into Checkpointer

SamTov committed:
Merge branch 'master' into Checkpointer

jdrotleff committed:
Update and rename checkpointer.py to base_checkpointer.py

Changing the name of the parent to BaseCheckpointer to make it more clear.

jdrotleff committed:
Update best_checkpointer.py

Added docstrings + compatibility with BaseCheckpointer

jdrotleff committed:
Update backup_checkpointer.py

Added docstrings + compatibility with BaseCheckpointer

jdrotleff committed:
Update goal_checkpointer.py

Add docstrings + compatibility with BaseCheckpointer

jdrotleff committed:
Update regular_checkpointer.py

Added docstrings + Compatibility with BaseCheckpointer

jdrotleff committed:
Update __init__.py

Changed Checkpointer to BaseCheckpointer

jdrotleff committed:
Update trainer.py

Changed Checkpointer to BaseCheckpointer

jdrotleff committed:
Create test_backup_checkpointer.py

Added a test for the backup checkpointer.

jdrotleff committed:
Create test_best_checkpointer.py

Added a test for the best checkpointer

jdrotleff committed:
Create test_regular_checkpointer.py

Add a test for the regular checkpointer module

jdrotleff committed:
Create test_goal_checkpointer.py

Added a test for the goal checkpointer

jdrotleff committed:
Fixed a bug in best_checkpointer.py

As the variable current episode starts at 0, a +1 is needed to stop at the correct episode.

jdrotleff committed:
Update backup_checkpointer.py

Fixed a bug where the change of the reward was not evaluated as  a checkpoint criterion.

Changed analyzed reward to an average reward instead of a single episode reward.

SamTov committed:
Merge branch 'master' into Checkpointer

jdrotleff committed:
Added checkpointing storage management

Each checkpointer now needs an out_path that sets where model checkpoints are stored.

jdrotleff committed:
Checkpointer Cleanup

jdrotleff committed:
fixed training break in goal_checkpointer.py

jdrotleff committed:
Initialize old_max to handle negative rewards.

SamTov committed:
Merge branch 'master' into Checkpointer

jdrotleff committed:
Trigger CI rerun

jdrotleff committed:
Adding typer and click to requirements

Attempt to solve the pytest error when installing dependencies:
ERROR: pip's dependency resolver does not currently take into account all the packages that are installed. This behaviour is the source of the following dependency conflicts.
typer 0.24.1 requires click>=8.2.1, but you have click 8.1.7 which is incompatible.

jdrotleff committed:
Removing click, limiting typer version to 0.12.0

jdrotleff committed:
Refactoring, validation, input guards

- Add input validation to `checkpointer.check_for_checkpoint()`
- add window_width and running_out_length validation
- add get_stop_episode to BaseCheckpointer (returns -1)
- Removed unused logger imports
- Changed rewards list to array for episodic and continuous trainer
- removed all print statements and turned them into
  logger.info/debug/warning
- off by one index error fix for episodic training break-timing

jdrotleff committed:
Merge branch 'master' into Checkpointer

jdrotleff committed:
pre-commit, formatting fix

jdrotleff committed:
Fixed wca typo

SamTov committed:
Merge branch 'master' into Checkpointer

SamTov committed:
Merge branch 'master' into Checkpointer

jdrotleff commented on swarmrl/checkpointers/backup_checkpointer.py:
```diff
@@ -0,0 +1,101 @@
+from swarmrl.checkpointers.base_checkpointer import BaseCheckpointer
+
+
+class BackupCheckpointer(BaseCheckpointer):
```
The idea was to save a checkpoint when forgetting occurs. So it saves a checkpoint if the performance drops after a reward peak, but is still above some minimum threshold (to not save a lot of noise at low rewards). This threshold is updated to the current average reward when a checkpoint is stored, so that the next checkpoint made should be a model with a better reward.
We could rename it for clarity.
```suggestion
class OnDeclineCheckpointer(BaseCheckpointer):
```
Alternatively, see the answer to your next comment

jdrotleff commented on swarmrl/checkpointers/best_checkpointer.py:
```diff
@@ -0,0 +1,86 @@
+from swarmrl.checkpointers.base_checkpointer import BaseCheckpointer
+
+
+class BestRewardCheckpointer(BaseCheckpointer):
```
Performance is different, yes.
We can remove the `BackupCheckpointer`and add a flag `only_save_on_decline` here. 

Then we might need to think about keeping a copy of the previous episode's model, as it was performing better than the current one, if reward is dropping again, or do we?

SamTov replied:
After discussion, I am happy with this.

jdrotleff commented on swarmrl/checkpointers/base_checkpointer.py:
Currently in the trainer. I saw your comment there. Thanks for pointing it out! I'll see how it can be done

jdrotleff commented on swarmrl/trainers/continuous_trainer.py:
```diff
@@ -64,26 +69,49 @@ def perform_rl_training(
+                if len(self.checkpointers) > 0:
+                    export = []
+                    save_string = ""
+                    for checkpointer in self.checkpointers:
```
I think we should add a `CheckpointManager` class that uses a callback to the export method of the trainer, manages the `save_string`logic and all active checkpointers.

SamTov replied:
What about the parent class?

jdrotleff replied:
Addressed it

jdrotleff committed:
Add `CheckpointManager` for actual checkpointing

- moved checkpointing managing into the CheckpointManager class, removed
  from the trainer
- added a `maybe_save_checkpoint`-method to the trainer and initializing
  the `CheckpointManager` in the init.
- Added unit test for `CheckpointManager`

jdrotleff committed:
Renaming backup checkpointer to on-decline checkpointer for clarity

jdrotleff committed:
move save logic into checkpointers, buffer added

- checkpointers now save via save_callback, manager only organizes
- add per-checkpointer retention buffer (n_buffer) with guarded cleanup
- improve base checkpointer internal naming and tests
- tempDir for `test_flax_network.py`

jdrotleff committed:
Changed default values for n_buffer

jdrotleff committed:
Fix: Allowed None out_path for individual checkpointer

jdrotleff committed:
Unify checkpointer logging and centralize training break handling

- switch checkpointer logging to Loguru for consistency
- move break decision logic into CheckpointManager
- add checkpointer break utility to ContinuousTrainer

jdrotleff committed:
Addded `save_only_on_first_goal_hit` option to goal checkpointer

- By default the goalcheckpointer does not checkpoint anymore after
  storing the first reached goal state, prevents continuous saving, when
no goal break is active
- added a finalize to continous_trainer.py to clean up the trajectory
  storing
- Unified episode displaying. Start with 1, end with n_episodes.

SamTov commented on swarmrl/trainers/trainer.py:
```diff
@@ -115,9 +151,54 @@ def export_models(self, directory: str = "Models"):
+            current_reward=current_reward,
+        )
+
+    def maybe_stop_training(self) -> tuple[bool, int]:
```
I do love this name. But perhaps we can find something a bit more technical like `check_for_stop_criterion`?

jdrotleff committed:
Rename `maybe_stop_training` to `check_for_stop_criterion`

- also edited docstring for `GoalCheckpointer`

SamTov reviewed (approved):
LGTM
